File: experiments-local/experiment-17/runs/agent=omp_language=python_model=openrouter/moonshotai/kimi-k2.7-code_tooling=none/rep1-failed/queries.py
"""Query layer for the Brazilian soccer knowledge graph.

All functions accept a `DataStore` instance and return plain Python data
structures so the answers are easy to turn into MCP tool responses.
"""

from __future__ import annotations

from datetime import date, datetime
from typing import Any

import pandas as pd

from data_loader import DataStore
from normalization import normalize_team_name, normalize_text, resolve_team_query


def _to_date(value: str | date | datetime | None) -> date | None:
    if value is None:
        return None
    if isinstance(value, datetime):
        return value.date()
    if isinstance(value, date):
        return value
    parsed = pd.to_datetime(value, errors="coerce", dayfirst="/" in str(value))
    return parsed.date() if pd.notna(parsed) else None


def _matching_canonicals(df: pd.DataFrame, query: str | None) -> set[str]:
    """Resolve a free-text team query to canonical team names in the data."""
    if not query:
        return set()
    q = str(query).strip()
    # Prefer the curated alias list first.
    matches = set(resolve_team_query(q))
    if len(matches) >= 1:
        return matches

    # Fallback: search all canonical names currently in the dataset.
    names = set(df["home_team"].unique()) | set(df["away_team"].unique())
    key = normalize_text(q)
    for name in names:
        nkey = normalize_text(name)
        if key == nkey or key in nkey:
            matches.add(name)
    return matches


def find_matches(
    ds: DataStore,
    team: str | None = None,
    opponent: str | None = None,
    competition: str | None = None,
    season: int | None = None,
    date_from: str | date | None = None,
    date_to: str | date | None = None,
    limit: int = 20,
) -> list[dict[str, Any]]:
    """Return matches matching the supplied filters."""
    df = ds.matches.copy()

    if competition:
        df = df[df["competition"].str.contains(competition, case=False, na=False)]

    if season:
        df = df[df["season"] == int(season)]

    d0 = _to_date(date_from)
    d1 = _to_date(date_to)
    if d0:
        df = df[df["date"] >= pd.Timestamp(d0)]
    if d1:
        df = df[df["date"] <= pd.Timestamp(d1)]

    team_names = _matching_canonicals(df, team) if team else set()
    opp_names = _matching_canonicals(df, opponent) if opponent else set()

    if team and not team_names:
        return []
    if opponent and not opp_names:
        return []

    if team_names:
        df = df[df["home_team"].isin(team_names) | df["away_team"].isin(team_names)]
    if opp_names:
        df = df[df["home_team"].isin(opp_names) | df["away_team"].isin(opp_names)]

    if team_names and opp_names:
        df = df[
            (df["home_team"].isin(team_names) | df["away_team"].isin(team_names))
            & (df["home_team"].isin(opp_names) | df["away_team"].isin(opp_names))
        ]

    df = df.sort_values("date", ascending=False).head(limit)
    return [_match_record(r) for _, r in df.iterrows()]


def _match_record(r: pd.Series) -> dict[str, Any]:
    return {
        "date": r["date"].strftime("%Y-%m-%d") if pd.notna(r["date"]) else None,
        "competition": r["competition"],
        "season": int(r["season"]) if pd.notna(r["season"]) else None,
        "home_team": r["home_team"],
        "away_team": r["away_team"],
        "home_goal": int(r["home_goal"]),
        "away_goal": int(r["away_goal"]),
        "round": r["round"] if pd.notna(r["round"]) else None,
        "stage": r["stage"] if pd.notna(r["stage"]) else None,
    }


def team_stats(
    ds: DataStore,
    team: str,
    season: int | None = None,
    competition: str | None = None,
    venue: str | None = None,
) -> dict[str, Any]:
    """Return win/loss/draw and goal statistics for a team."""
    df = ds.matches.copy()
    teams = _matching_canonicals(df, team)
    if not teams:
        return {"team": team, "matches": 0, "error": "Team not found"}

    if competition:
        df = df[df["competition"].str.contains(competition, case=False, na=False)]
    if season:
        df = df[df["season"] == int(season)]

    if venue and venue.lower() == "home":
        df = df[df["home_team"].isin(teams)]
    elif venue and venue.lower() == "away":
        df = df[df["away_team"].isin(teams)]
    else:
        df = df[df["home_team"].isin(teams) | df["away_team"].isin(teams)]

    wins = draws = losses = gf = ga = 0
    for _, r in df.iterrows():
        is_home = r["home_team"] in teams
        team_goals = r["home_goal"] if is_home else r["away_goal"]
        opp_goals = r["away_goal"] if is_home else r["home_goal"]
        gf += int(team_goals)
        ga += int(opp_goals)
        if team_goals > opp_goals:
            wins += 1
        elif team_goals < opp_goals:
            losses += 1
        else:
            draws += 1

    matches = wins + draws + losses
    return {
        "team": team,
        "canonical_teams": sorted(teams),
        "season": season,
        "competition": competition,
        "venue": venue,
        "matches": matches,
        "wins": wins,
        "draws": draws,
        "losses": losses,
        "goals_for": gf,
        "goals_against": ga,
        "goal_difference": gf - ga,
        "win_rate": round(wins / matches * 100, 1) if matches else 0.0,
    }


def head_to_head(
    ds: DataStore,
    team_a: str,
    team_b: str,
    competition: str | None = None,
    season: int | None = None,
    limit: int = 50,
) -> dict[str, Any]:
    """Return all matches between two teams plus a summary."""
    df = ds.matches.copy()
    a_teams = _matching_canonicals(df, team_a)
    b_teams = _matching_canonicals(df, team_b)

    df = df[
        (df["home_team"].isin(a_teams) | df["away_team"].isin(a_teams))
        & (df["home_team"].isin(b_teams) | df["away_team"].isin(b_teams))
    ]

    if competition:
        df = df[df["competition"].str.contains(competition, case=False, na=False)]
    if season:
        df = df[df["season"] == int(season)]

    a_wins = b_wins = draws = a_gf = b_gf = 0
    matches_all = []
    for _, r in df.sort_values("date", ascending=False).iterrows():
        rec = _match_record(r)
        matches_all.append(rec)
        h_in_a = r["home_team"] in a_teams
        a_in_a = r["away_team"] in a_teams
        hg = int(r["home_goal"])
        ag = int(r["away_goal"])
        if h_in_a:
            a_gf += hg
            b_gf += ag
        else:
            a_gf += ag
            b_gf += hg

        if hg == ag:
            draws += 1
        elif h_in_a and hg > ag:
            a_wins += 1
        elif a_in_a and ag > hg:
            a_wins += 1
        else:
            b_wins += 1

    return {
        "team_a": team_a,
        "team_b": team_b,
        "matches": len(matches_all),
        "team_a_wins": a_wins,
        "team_b_wins": b_wins,
        "draws": draws,
        "team_a_goals": a_gf,
        "team_b_goals": b_gf,
        "match_list": matches_all[:limit],
    }


def biggest_wins(
    ds: DataStore,
    competition: str | None = None,
    season: int | None = None,
    limit: int = 10,
) -> list[dict[str, Any]]:
    """Return the biggest wins (largest goal difference) in the dataset."""
    df = ds.matches.copy()
    if competition:
        df = df[df["competition"].str.contains(competition, case=False, na=False)]
    if season:
        df = df[df["season"] == int(season)]
    df = df.dropna(subset=["home_goal", "away_goal"])
    df["margin"] = (df["home_goal"] - df["away_goal"]).abs()
    df = df.sort_values(["margin", "date"], ascending=[False, False]).head(limit)
    return [_match_record(r) for _, r in df.iterrows()]


def average_goals_per_match(
    ds: DataStore, competition: str | None = None, season: int | None = None
) -> dict[str, float]:
    """Average total goals per match and home win rate."""
    df = ds.matches.copy()
    if competition:
        df = df[df["competition"].str.contains(competition, case=False, na=False)]
    if season:
        df = df[df["season"] == int(season)]
    df = df.dropna(subset=["home_goal", "away_goal"])
    total = len(df)
    if total == 0:
        return {"matches": 0, "average_goals": 0.0, "home_win_rate": 0.0}
    avg = (df["home_goal"] + df["away_goal"]).mean()
    home_wins = (df["home_goal"] > df["away_goal"]).sum()
    return {
        "matches": int(total),
        "average_goals": round(float(avg), 2),
        "home_win_rate": round(float(home_wins / total * 100), 1),
    }


def best_home_record(ds: DataStore, competition: str | None = None) -> dict[str, Any]:
    """Return the team with the best home record (by win rate, min 5 home games)."""
    df = ds.matches.copy()
    if competition:
        df = df[df["competition"].str.contains(competition, case=False, na=False)]
    df = df.dropna(subset=["home_goal", "away_goal"])
    home = df.groupby("home_team").apply(
        lambda g: pd.Series(
            {
                "matches": len(g),
                "wins": int((g["home_goal"] > g["away_goal"]).sum()),
                "draws": int((g["home_goal"] == g["away_goal"]).sum()),
                "losses": int((g["home_goal"] < g["away_goal"]).sum()),
                "goals_for": int(g["home_goal"].sum()),
                "goals_against": int(g["away_goal"].sum()),
            }
        )
    )
    home = home[home["matches"] >= 5].copy()
    home["win_rate"] = home["wins"] / home["matches"]
    home = home.sort_values(["win_rate", "goals_for"], ascending=[False, False])
    if home.empty:
        return {"team": None, "matches": 0, "win_rate": 0.0}
    top = home.iloc[0]
    return {
        "team": top.name,
        "matches": int(top["matches"]),
        "wins": int(top["wins"]),
        "draws": int(top["draws"]),
        "losses": int(top["losses"]),
        "goals_for": int(top["goals_for"]),
        "goals_against": int(top["goals_against"]),
        "win_rate": round(float(top["win_rate"]) * 100, 1),
    }


def search_players(
    ds: DataStore,
    name: str | None = None,
    nationality: str | None = None,
    club: str | None = None,
    position: str | None = None,
    min_overall: int | None = None,
    max_overall: int | None = None,
    limit: int = 20,
) -> list[dict[str, Any]]:
    """Search the FIFA player dataset."""
    df = ds.players.copy()
    if name:
        df = df[df["name"].str.contains(name, case=False, na=False)]
    if nationality:
        df = df[df["nationality"].str.contains(nationality, case=False, na=False)]
    if club:
        df = df[df["club"].str.contains(club, case=False, na=False)]
    if position:
        df = df[df["position"].str.contains(position, case=False, na=False)]
    if min_overall is not None:
        df = df[df["overall"] >= int(min_overall)]
    if max_overall is not None:
        df = df[df["overall"] <= int(max_overall)]
    df = df.sort_values("overall", ascending=False).head(limit)
    return [_player_record(r) for _, r in df.iterrows()]


def top_players(
    ds: DataStore,
    nationality: str | None = None,
    club: str | None = None,
    position: str | None = None,
    limit: int = 10,
) -> list[dict[str, Any]]:
    """Convenience wrapper for the highest-rated players by criteria."""
    return search_players(
        ds,
        nationality=nationality,
        club=club,
        position=position,
        limit=limit,
    )


def _player_record(r: pd.Series) -> dict[str, Any]:
    return {
        "id": int(r["id"]) if pd.notna(r["id"]) else None,
        "name": r["name"],
        "age": int(r["age"]) if pd.notna(r["age"]) else None,
        "nationality": r["nationality"],
        "overall": int(r["overall"]) if pd.notna(r["overall"]) else None,
        "potential": int(r["potential"]) if pd.notna(r["potential"]) else None,
        "club": r["club"],
        "position": r["position"],
        "jersey_number": int(r["jersey_number"]) if pd.notna(r["jersey_number"]) else None,
        "height": r["height"],
        "weight": r["weight"],
    }


def team_competitions(ds: DataStore, team: str) -> list[dict[str, Any]]:
    """List competitions and seasons a team has played in."""
    df = ds.matches.copy()
    teams = _matching_canonicals(df, team)
    if not teams:
        return []
    df = df[df["home_team"].isin(teams) | df["away_team"].isin(teams)]
    grouped = (
        df.groupby(["competition", "season"], dropna=False)
        .size()
        .reset_index(name="matches")
    )
    return [
        {"competition": r["competition"], "season": int(r["season"]), "matches": int(r["matches"])}
        for _, r in grouped.iterrows()
    ]


def list_competitions(ds: DataStore) -> list[str]:
    """Return all competition names in the match data."""
    return sorted(ds.matches["competition"].dropna().unique().tolist())


def list_seasons(ds: DataStore, competition: str | None = None) -> list[int]:
    """Return all seasons, optionally filtered to a competition."""
    df = ds.matches
    if competition:
        df = df[df["competition"].str.contains(competition, case=False, na=False)]
    return sorted([int(s) for s in df["season"].dropna().unique()])

File: experiments-local/experiment-mu-glm52-ompfix/runs/agent=omp_language=typescript_model=openrouter/z-ai/glm-5.2_tooling=none/rep2/src/queries.ts
/**
 * Brazilian Soccer MCP Server — query engine
 * ------------------------------------------
 * Context block:
 *   Pure functions over the loaded Dataset that implement the five query
 *   categories required by the spec: match, team, player, competition and
 *   statistical analysis. Results are returned as plain JSON-serialisable
 *   objects so the MCP tool layer can serialise them verbatim.
 *
 *   Team matching uses the canonical team key (normalizer.canonicalTeamKey)
 *   so a query for "Flamengo" matches "Flamengo-RJ", "Flamengo - RJ" and
 *   "Flamengo" across files. A fuzzy fallback (substring on display name)
 *   is offered when an exact canonical lookup yields nothing, so the user can
 *   ask about teams whose exact alias is not curated.
 */

import type {
  HeadToHead,
  Match,
  StandingRow,
  TeamStats,
} from "./types.js";
import { canonicalTeamKey, teamKey } from "./normalizer.js";
import { formatDate } from "./dates.js";
import type { Dataset } from "./loader.js";

export interface MatchQuery {
  team?: string;
  opponent?: string;
  homeTeam?: string;
  awayTeam?: string;
  competition?: string;
  season?: number;
  startDate?: string;
  endDate?: string;
  stage?: string;
  round?: string;
  limit?: number;
}

export interface MatchResult {
  id: string;
  date: string;
  competition: string;
  homeTeam: string;
  awayTeam: string;
  homeGoal: number | null;
  awayGoal: number | null;
  season: number | null;
  round: string | null;
  stage: string | null;
  arena: string | null;
  source: string;
}

export interface TeamQuery {
  team: string;
  season?: number;
  competition?: string;
  homeAway?: "home" | "away" | "all";
}

export interface PlayerQuery {
  name?: string;
  nationality?: string;
  club?: string;
  position?: string;
  minOverall?: number;
  sortBy?: "overall" | "potential" | "age" | "name";
  limit?: number;
}

export interface CompetitionQuery {
  competition?: string;
  season?: number;
}

export interface StatsQuery {
  competition?: string;
  season?: number;
  team?: string;
}


/** Does a match's competition label match the requested competition? */
function competitionMatches(match: Match, requested?: string): boolean {
  if (!requested) return true;
  const want = requested.toLowerCase();
  return match.competition.toLowerCase().includes(want);
}

/** Convert a Match to the user-facing result shape. */
function toResult(m: Match): MatchResult {
  return {
    id: m.id,
    date: formatDate(m.date, m.rawDate),
    competition: m.competition,
    homeTeam: m.homeTeamDisplay,
    awayTeam: m.awayTeamDisplay,
    homeGoal: m.homeGoal,
    awayGoal: m.awayGoal,
    season: m.season,
    round: m.round,
    stage: m.stage ?? null,
    arena: m.arena ?? null,
    source: m.source,
  };
}

/** Parse an ISO date string into a UTC midnight Date, or null. */
function parseIsoDate(raw: string): Date | null {
  const d = new Date(raw);
  if (isNaN(d.getTime())) return null;
  return new Date(Date.UTC(d.getUTCFullYear(), d.getUTCMonth(), d.getUTCDate()));
}

/** Query matches by criteria. */
export function queryMatches(ds: Dataset, q: MatchQuery): MatchResult[] {
  const teamKeyQ = q.team ? canonicalTeamKey(q.team) : undefined;
  const oppKeyQ = q.opponent ? canonicalTeamKey(q.opponent) : undefined;
  const homeKeyQ = q.homeTeam ? canonicalTeamKey(q.homeTeam) : undefined;
  const awayKeyQ = q.awayTeam ? canonicalTeamKey(q.awayTeam) : undefined;
  const start = q.startDate ? parseIsoDate(q.startDate) : null;
  const end = q.endDate ? parseIsoDate(q.endDate) : null;

  const out: MatchResult[] = [];
  for (const m of ds.matches) {
    if (teamKeyQ && m.homeTeam !== teamKeyQ && m.awayTeam !== teamKeyQ) continue;
    if (oppKeyQ) {
      const otherIsOpp = (m.homeTeam === teamKeyQ && m.awayTeam === oppKeyQ) ||
        (m.awayTeam === teamKeyQ && m.homeTeam === oppKeyQ);
      if (!otherIsOpp) continue;
    }
    if (homeKeyQ && m.homeTeam !== homeKeyQ) continue;
    if (awayKeyQ && m.awayTeam !== awayKeyQ) continue;
    if (!competitionMatches(m, q.competition)) continue;
    if (q.season !== undefined && m.season !== q.season) continue;
    if (q.round && m.round !== q.round) continue;
    if (q.stage && m.stage && !m.stage.toLowerCase().includes(q.stage.toLowerCase())) continue;
    if (start && m.date && m.date < start) continue;
    if (end && m.date && m.date > end) continue;
    out.push(toResult(m));
  }

  out.sort((a, b) => (b.date > a.date ? 1 : b.date < a.date ? -1 : 0));
  if (q.limit && q.limit > 0) return out.slice(0, q.limit);
  return out;
}

/** Compute aggregate statistics for a team, optionally filtered. */
export function teamStats(ds: Dataset, q: TeamQuery): TeamStats {
  const key = canonicalTeamKey(q.team);
  let wins = 0, draws = 0, losses = 0, goalsFor = 0, goalsAgainst = 0;
  let teamDisplay = q.team;
  let count = 0;

  for (const m of ds.matches) {
    const isHome = m.homeTeam === key;
    const isAway = m.awayTeam === key;
    if (!isHome && !isAway) continue;
    if (q.season !== undefined && m.season !== q.season) continue;
    if (!competitionMatches(m, q.competition)) continue;
    if (q.homeAway === "home" && !isHome) continue;
    if (q.homeAway === "away" && !isAway) continue;
    if (teamDisplay === q.team) teamDisplay = isHome ? m.homeTeamDisplay : m.awayTeamDisplay;

    const hg = m.homeGoal ?? 0;
    const ag = m.awayGoal ?? 0;
    if (isHome) {
      goalsFor += hg;
      goalsAgainst += ag;
      if (hg > ag) wins++;
      else if (hg < ag) losses++;
      else draws++;
    } else {
      goalsFor += ag;
      goalsAgainst += hg;
      if (ag > hg) wins++;
      else if (ag < hg) losses++;
      else draws++;
    }
    count++;
  }

  return {
    team: key,
    teamDisplay,
    matches: count,
    wins,
    draws,
    losses,
    goalsFor,
    goalsAgainst,
    points: wins * 3 + draws,
    homeAway: q.homeAway ?? "all",
  };
}

/** Head-to-head summary between two teams. */
export function headToHead(ds: Dataset, teamA: string, teamB: string, season?: number): HeadToHead {
  const keyA = canonicalTeamKey(teamA);
  const keyB = canonicalTeamKey(teamB);
  let aWins = 0, bWins = 0, draws = 0, aGoals = 0, bGoals = 0, matches = 0;

  for (const m of ds.matches) {
    if (season !== undefined && m.season !== season) continue;
    const aIsHome = m.homeTeam === keyA && m.awayTeam === keyB;
    const aIsAway = m.awayTeam === keyA && m.homeTeam === keyB;
    if (!aIsHome && !aIsAway) continue;
    const hg = m.homeGoal ?? 0;
    const ag = m.awayGoal ?? 0;
    if (aIsHome) {
      aGoals += hg;
      bGoals += ag;
      if (hg > ag) aWins++;
      else if (hg < ag) bWins++;
      else draws++;
    } else {
      aGoals += ag;
      bGoals += hg;
      if (ag > hg) aWins++;
      else if (hg > ag) bWins++;
      else draws++;
    }
    matches++;
  }

  return {
    teamA: keyA,
    teamB: keyB,
    matches,
    teamAWins: aWins,
    teamBWins: bWins,
    draws,
    teamAGoals: aGoals,
    teamBGoals: bGoals,
  };
}

/** Query FIFA players by criteria. */
export function queryPlayers(ds: Dataset, q: PlayerQuery) {
  const nameQ = q.name?.toLowerCase();
  const natQ = q.nationality?.toLowerCase();
  const clubQ = q.club?.toLowerCase();
  const posQ = q.position?.toLowerCase();

  let rows = ds.players.filter((p) => {
    if (nameQ && !p.name.toLowerCase().includes(nameQ)) return false;
    if (natQ && !p.nationality.toLowerCase().includes(natQ)) return false;
    if (clubQ && !p.club.toLowerCase().includes(clubQ)) return false;
    if (posQ && !p.position.toLowerCase().includes(posQ)) return false;
    if (q.minOverall !== undefined && (p.overall ?? 0) < q.minOverall) return false;
    return true;
  });

  const sortBy = q.sortBy ?? "overall";
  rows = rows.slice().sort((a, b) => {
    switch (sortBy) {
      case "overall":
        return (b.overall ?? 0) - (a.overall ?? 0);
      case "potential":
        return (b.potential ?? 0) - (a.potential ?? 0);
      case "age":
        return (a.age ?? 0) - (b.age ?? 0);
      case "name":
        return a.name.localeCompare(b.name);
      default:
        return 0;
    }
  });

  if (q.limit && q.limit > 0) rows = rows.slice(0, q.limit);
  return rows;
}

/** Standings for a competition+season, computed from match results (3-1-0). */
export function standings(ds: Dataset, q: CompetitionQuery): StandingRow[] {
  const table = new Map<string, {
    team: string; display: string; played: number; w: number; d: number; l: number;
    gf: number; ga: number;
  }>();

  for (const m of ds.matches) {
    if (!competitionMatches(m, q.competition)) continue;
    if (q.season !== undefined && m.season !== q.season) continue;
    if (m.homeGoal === null || m.awayGoal === null) continue;

    ensureRow(table, m.homeTeam, m.homeTeamDisplay);
    ensureRow(table, m.awayTeam, m.awayTeamDisplay);
    const home = table.get(m.homeTeam)!;
    const away = table.get(m.awayTeam)!;
    home.played++; away.played++;
    home.gf += m.homeGoal; home.ga += m.awayGoal;
    away.gf += m.awayGoal; away.ga += m.homeGoal;
    if (m.homeGoal > m.awayGoal) { home.w++; away.l++; }
    else if (m.homeGoal < m.awayGoal) { away.w++; home.l++; }
    else { home.d++; away.d++; }
  }

  return Array.from(table.values())
    .map((r) => ({
      position: 0,
      team: r.team,
      teamDisplay: r.display,
      played: r.played,
      wins: r.w,
      draws: r.d,
      losses: r.l,
      goalsFor: r.gf,
      goalsAgainst: r.ga,
      goalDifference: r.gf - r.ga,
      points: r.w * 3 + r.d,
    }))
    .sort((a, b) => b.points - a.points || b.goalDifference - a.goalDifference || b.goalsFor - a.goalsFor)
    .map((r, i) => ({ ...r, position: i + 1 }));
}

function ensureRow(
  table: Map<string, { team: string; display: string; played: number; w: number; d: number; l: number; gf: number; ga: number }>,
  key: string,
  display: string,
): void {
  if (!table.has(key)) {
    table.set(key, { team: key, display, played: 0, w: 0, d: 0, l: 0, gf: 0, ga: 0 });
  } else {
    const existing = table.get(key)!;
    if (!existing.display) existing.display = display;
  }
}

/** Average goals per match, home win rate, away win rate, draw rate. */
export function goalStats(ds: Dataset, q: StatsQuery) {
  let totalGoals = 0, totalMatches = 0;
  let homeWins = 0, awayWins = 0, draws = 0;
  let biggestWins: { date: string; home: string; away: string; homeGoal: number; awayGoal: number; competition: string; margin: number }[] = [];
  const teamKeyQ = q.team ? canonicalTeamKey(q.team) : undefined;

  for (const m of ds.matches) {
    if (!competitionMatches(m, q.competition)) continue;
    if (q.season !== undefined && m.season !== q.season) continue;
    if (m.homeGoal === null || m.awayGoal === null) continue;
    if (teamKeyQ && m.homeTeam !== teamKeyQ && m.awayTeam !== teamKeyQ) continue;

    totalGoals += m.homeGoal + m.awayGoal;
    totalMatches++;
    if (m.homeGoal > m.awayGoal) homeWins++;
    else if (m.awayGoal > m.homeGoal) awayWins++;
    else draws++;

    const margin = Math.abs(m.homeGoal - m.awayGoal);
    if (margin >= 4) {
      biggestWins.push({
        date: formatDate(m.date, m.rawDate),
        home: m.homeTeamDisplay,
        away: m.awayTeamDisplay,
        homeGoal: m.homeGoal,
        awayGoal: m.awayGoal,
        competition: m.competition,
        margin,
      });
    }
  }

  biggestWins.sort((a, b) => b.margin - a.margin || b.homeGoal + b.awayGoal - (a.homeGoal + a.awayGoal));
  return {
    totalMatches,
    totalGoals,
    averageGoalsPerMatch: totalMatches ? +(totalGoals / totalMatches).toFixed(2) : 0,
    homeWinRate: totalMatches ? +((homeWins / totalMatches) * 100).toFixed(1) : 0,
    awayWinRate: totalMatches ? +((awayWins / totalMatches) * 100).toFixed(1) : 0,
    drawRate: totalMatches ? +((draws / totalMatches) * 100).toFixed(1) : 0,
    biggestWins: biggestWins.slice(0, 10),
  };
}

/** Best home or away record across a competition/season. */
export function bestRecord(ds: Dataset, q: { competition?: string; season?: number; homeAway: "home" | "away" }): TeamStats[] {
  const teams = distinctTeams(ds, q);
  const stats: TeamStats[] = teams.map((t) => teamStats(ds, {
    team: t.display,
    competition: q.competition,
    season: q.season,
    homeAway: q.homeAway,
  }));
  return stats
    .filter((s) => s.matches > 0)
    .sort((a, b) => b.points - a.points || (b.wins - b.losses) - (a.wins - a.losses) || b.goalsFor - a.goalsFor)
    .slice(0, 5);
}

function distinctTeams(ds: Dataset, q: { competition?: string; season?: number }): { key: string; display: string }[] {
  const map = new Map<string, string>();
  for (const m of ds.matches) {
    if (!competitionMatches(m, q.competition)) continue;
    if (q.season !== undefined && m.season !== q.season) continue;
    if (!map.has(m.homeTeam)) map.set(m.homeTeam, m.homeTeamDisplay);
    if (!map.has(m.awayTeam)) map.set(m.awayTeam, m.awayTeamDisplay);
  }
  return Array.from(map.entries()).map(([key, display]) => ({ key, display }));
}

/** Top scorers are not directly present; we approximate by goals per team-game. */
export function topScoringTeams(ds: Dataset, q: { competition?: string; season?: number; limit?: number }) {
  const stats: { team: string; display: string; goalsFor: number; matches: number; avg: number }[] = [];
  const teams = distinctTeams(ds, q);
  for (const t of teams) {
    const s = teamStats(ds, { team: t.display, competition: q.competition, season: q.season });
    if (s.matches === 0) continue;
    stats.push({
      team: s.team,
      display: s.teamDisplay,
      goalsFor: s.goalsFor,
      matches: s.matches,
      avg: +(s.goalsFor / s.matches).toFixed(2),
    });
  }
  stats.sort((a, b) => b.goalsFor - a.goalsFor || b.avg - a.avg);
  return stats.slice(0, q.limit ?? 10);
}
/** Resolve a free-text team name to canonical info (used by MCP tools). */
export function resolveTeam(ds: Dataset, name: string) {
  const key = canonicalTeamKey(name);
  let display = name;
  for (const m of ds.matches) {
    if (m.homeTeam === key) { display = m.homeTeamDisplay; break; }
    if (m.awayTeam === key) { display = m.awayTeamDisplay; break; }
  }
  // Fuzzy fallback: substring on the deaccented display name when no exact
  // canonical-key match exists. Returns whichever side actually matched.
  if (!ds.matches.some((m) => m.homeTeam === key || m.awayTeam === key)) {
    const needle = teamKey(name);
    if (needle) {
      for (const m of ds.matches) {
        const homeHit = teamKey(m.homeTeamDisplay).includes(needle);
        const awayHit = teamKey(m.awayTeamDisplay).includes(needle);
        if (homeHit || awayHit) {
          return {
            query: name,
            canonicalKey: homeHit ? m.homeTeam : m.awayTeam,
            matchedDisplay: homeHit ? m.homeTeamDisplay : m.awayTeamDisplay,
            fuzzy: true,
          };
        }
      }
    }
  }
  return { query: name, canonicalKey: key, matchedDisplay: display, fuzzy: false };
}
